tools/prefetch_cache_gcsfuse/prefetch.go

```diff
@@ -0,0 +1,142 @@
+		BucketName:     object.Bucket,
+		ObjectName:     object.Name,
+		Generation:     object.Generation,
+		MetaGeneration: object.Metageneration,
+	}
+
+	// TODO ezl refactor once we decouple content cache from gcsx
+	// Create a temporary cache file on disk
+	// TODO ezl concerns about partial download and cache file corruption
+	// We may want a way to verify the files are fully downloaded
+	// and either resuming the download or discarding and redownloading the file
+	f, err := ioutil.TempFile(cacheDir, gcsx.CACHE_FILE_PREFIX)
+
+	if err != nil {
+		err = fmt.Errorf("ioutil.TempFile: %w", err)
+		return
+	}
+
+	rc, err := client.Bucket(object.Bucket).Object(object.Name).NewReader(ctx)
+
+	if err != nil {
+		return fmt.Errorf("Object(%q).NewReader: %v", object.Name, err)
+	}
+	defer rc.Close()
+
+	if _, err := io.Copy(f, rc); err != nil {
+		return fmt.Errorf("io.Copy: %v", err)
+	}
+
+	if err = f.Close(); err != nil {
+		return fmt.Errorf("f.Close: %v", err)
+	}
+
+	file, err := json.MarshalIndent(*tempFileObjectMetadata, "", " ")
+	err = ioutil.WriteFile(fmt.Sprintf("%s.json", f.Name()), file, 0644)
+	if err != nil {
+		err = fmt.Errorf("downloadFile failed to write metadata: %w", err)
+	}
+	return
+}
+
+func prefetchCache(cacheDir, bucketName, prefix, delimiter string) (err error) {
+	start := time.Now()
+
+	var wg sync.WaitGroup
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("storage.NewClient: %v", err)
+	}
+	defer client.Close()
+
+	// Prefixes and delimiters can be used to emulate directory listings.
+	// Prefixes can be used to filter objects starting with prefix.
+	// The delimiter argument can be used to restrict the results to only the
+	// objects in the given "directory". Without the delimiter, the entire tree
+	// under the prefix is returned.
+	//
+	// For example, given these blobs:
+	//   /a/1.txt
+	//   /a/b/2.txt
+	//
+	// If you just specify prefix="a/", you'll get back:
+	//   /a/1.txt
+	//   /a/b/2.txt
+	//
+	// However, if you specify prefix="a/" and delim="/", you'll get back:
+	//   /a/1.txt
+
+	// Should we set a higher timeout?
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*10)
+	defer cancel()
+
+	it := client.Bucket(bucketName).Objects(ctx, &storage.Query{
+		Prefix:    prefix,
+		Delimiter: delimiter,
+	})
+
+	// Concurrently download files from specified gcs bucket with prefix and delimiter
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("Bucket(%q).Objects: %v", bucketName, err)
+		}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err = downloadFile(ctx, client, attrs, cacheDir)
+			if err != nil {
+				log.Printf("prefetchCache: %v", err)
+			}
+		}()
+	}
+
+	// Wait for all goroutines downloading files to finish before returning
+	wg.Wait()
+
+	elapsed := time.Since(start)
+	log.Printf("Prefetch cache took %s", elapsed)
```

Comment:
It would be nice to print out stats about how many objects are attempted vs. downloaded.